src/enzyme_ad/jax/Passes/CanonicalizeLoops.cpp

```diff
@@ -614,26 +615,184 @@ struct CanonicalizeLoopsPass
           if (!constantRangeValue.has_value())
             return;
-          if (constantRangeValue->eq(cstRhs)) {
+          b.setInsertionPoint(cmpiOp);
+          auto cst = b.create<arith::ConstantOp>(
+              cmpiOp.getLoc(), b.getI1Type(),
+              IntegerAttr::get(b.getI1Type(), !constantRangeValue->eq(cstRhs)));
+          cmpiOp.getResult().replaceAllUsesWith(cst);
+        }
+        if (pred == arith::CmpIPredicate::eq) {
+          std::optional<APInt> constantRangeValue =
+              range.getValue().getConstantValue();
+          if (!constantRangeValue.has_value())
+            return;
+          b.setInsertionPoint(cmpiOp);
+          auto cst = b.create<arith::ConstantOp>(
+              cmpiOp.getLoc(), b.getI1Type(),
+              IntegerAttr::get(b.getI1Type(), constantRangeValue->eq(cstRhs)));
+          cmpiOp.getResult().replaceAllUsesWith(cst);
+        }
+        if (pred == arith::CmpIPredicate::ult) {
+          const APInt umax = cstRange.umax();
+          const APInt umin = cstRange.umin();
+          if (umax.ult(cstRhs)) {
+            // Condition always true.
+            b.setInsertionPoint(cmpiOp);
+            auto cst = b.create<arith::ConstantOp>(
+                cmpiOp.getLoc(), b.getI1Type(),
+                IntegerAttr::get(b.getI1Type(), true));
+            cmpiOp.getResult().replaceAllUsesWith(cst);
+          }
+          // range < cst -> !(range >= cst)
+          if (umin.uge(cstRhs)) {
+            // Condition always false.
             b.setInsertionPoint(cmpiOp);
             auto cst = b.create<arith::ConstantOp>(
                 cmpiOp.getLoc(), b.getI1Type(),
                 IntegerAttr::get(b.getI1Type(), false));
             cmpiOp.getResult().replaceAllUsesWith(cst);
           }
         }
-        if (pred == arith::CmpIPredicate::ult) {
+        if (pred == arith::CmpIPredicate::ule) {
+          const APInt umax = cstRange.umax();
+          const APInt umin = cstRange.umin();
+          if (umax.ule(cstRhs)) {
+            // Condition always true.
+            b.setInsertionPoint(cmpiOp);
+            auto cst = b.create<arith::ConstantOp>(
+                cmpiOp.getLoc(), b.getI1Type(),
+                IntegerAttr::get(b.getI1Type(), true));
+            cmpiOp.getResult().replaceAllUsesWith(cst);
+          }
+          // range <= cst -> !(range > cst)
+          if (umin.ugt(cstRhs)) {
+            // Condition always false.
+            b.setInsertionPoint(cmpiOp);
+            auto cst = b.create<arith::ConstantOp>(
+                cmpiOp.getLoc(), b.getI1Type(),
+                IntegerAttr::get(b.getI1Type(), false));
+            cmpiOp.getResult().replaceAllUsesWith(cst);
+          }
+        }
+        if (pred == arith::CmpIPredicate::ugt) {
           const APInt umax = cstRange.umax();
           const APInt umin = cstRange.umin();
-          if (umax.ult(cstRhs) && umin.ult(cstRhs)) {
+          if (umax.ugt(cstRhs)) {
```

Comment:
@wsmoses  I am not convinced. I think we should check the min not the max.

Reply:
Yeah I see what you mean

we should check against https://llvm.org/doxygen/ConstantRange_8cpp_source.html#l00243 for each of these

can you give it a go?

Reply:
Ok.